### ISSUE TITLE
cached_property return type is lost

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#449]
**Describe the bug**
Python 3.8 added a new descriptor decorator, [`functools.cached_property`](https://docs.python.org/3.8/library/functools.html#functools.cached_property), a cached variant of the built-in `property`.  The return type of methods decorated with `cached_property` is not kept:

```json
{
	"resource": "test.py",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 4,
	"message": "Type of 'Foo().bar' is 'None'",
	"source": "pyright",
	"startLineNumber": 14,
	"startColumn": 13,
	"endLineNumber": 14,
	"endColumn": 22
}
{
	"resource": "test.py",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 4,
	"message": "Type of 'Foo().baz' is 'Unknown'",
	"source": "pyright",
	"startLineNumber": 15,
	"startColumn": 13,
	"endLineNumber": 15,
	"endColumn": 22
}
```

**To Reproduce**
Type check the snippet below.

**Expected behavior**
The return type of `Foo().baz` should be `None`.

**Screenshots or Code**
```py
from functools import cached_property


class Foo:
    @property
    def bar(self) -> None:
        return None

    @cached_property
    def baz(self) -> None:
        return None


reveal_type(Foo().bar)
reveal_type(Foo().baz)
```

**VS Code extension or command-line**
extension, 1.1.12

**Comment by layday** [comment]
I also wonder how this might be generalised to user-defined descriptors, so that the undecorated return type is retained.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release of pyright.

Unfortunately, this doesn't help with user-defined decorators that are not properly annotated.

**Comment by erictraut** [comment]
This is now fixed in 1.1.13, which I just published.

**Comment by layday** [comment]
Confirmed that this also works with user-defined decorators which inherit from `Generic` - thanks a bunch.
